src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -187,19 +226,21 @@ public function setValue(&$objectOrArray, $propertyPath, $value)
     public static function handleError($type, $message, $file, $line, $context)
     {
         if (E_RECOVERABLE_ERROR === $type) {
-            self::throwUnexpectedTypeException($message, debug_backtrace(false), 1);
+            self::throwInvalidArgumentException($message, debug_backtrace(false), 1);
         }
 
         return null !== self::$previousErrorHandler && false !== call_user_func(self::$previousErrorHandler, $type, $message, $file, $line, $context);
     }
 
-    private static function throwUnexpectedTypeException($message, $trace, $i)
+    private static function throwInvalidArgumentException($message, $trace, $i)
     {
         if (isset($trace[$i]['file']) && __FILE__ === $trace[$i]['file']) {
             $pos = strpos($message, $delim = 'must be of the type ') ?: strpos($message, $delim = 'must be an instance of ');
             $pos += strlen($delim);
+            $type = $trace[$i]['args'][0];
+            $type = is_object($type) ? get_class($type) : gettype($type);
 
-            throw new UnexpectedTypeException($trace[$i]['args'][0], substr($message, $pos, strpos($message, ',', $pos) - $pos));
+            throw new InvalidArgumentException(sprintf('Expected argument of type "%s", "%s" given', substr($message, $pos, strpos($message, ',', $pos) - $pos), $type));
```

Comment:
Using UnexpectedTypeException in not compatible with 2.7, where this exception type has been specialized. Using InvalidArgumentException backported from 2.7 instead.